Fix #1578, use phonenumbers library for validation

floraline opened the pull request:
Fixes #1578.

This PR replaces the regex in `model_checks.py` for phone number validation with [this Python port](https://github.com/daviddrysdale/python-phonenumbers) of [Google's phone numbers library](https://github.com/googlei18n/libphonenumber). `phonenumbers` appears to already be accessible in this repo.

Input is parsed as a US phone number, unless a country code is provided. This library should handle international numbers correctly as long as the country code is known. Certain non-digit characters are acceptable, allowing for input like `(703) 123-4567`.

Note that I am using `is_possible_number` here instead of `is_valid_number`. The former only checks that the input is *like* a phone number (correct number of digits). The latter requires that the number is *valid* (a US number must have a valid area code).

I am merging the two validation failure messages here mostly because I don't know why there are two, especially now that we can validate international numbers. I also changed it from imperative to past tense, because this message only appears if the user has already entered something into this field.

We should be able to use this library to save inputted phone numbers in a normalized format like E.164.

floraline commented:
Whoops, looks like I missed something.

I see `ImportError: No module named 'phonenumbers'` in the travis build log. Maybe the `phonenumbers` library is not already installed in this project? Or did I not include it correctly? I don't know how to resolve this or where to start looking. It worked on my machine for whatever reason! 😅

Can someone please advise?

RobRuana commented:
Hi @floraline! `phonenumbers` is imported by the `tabletop` plugin, which is probably why it was working on your local deployment.

The test suites for each plugin run with the minimal requirements, so the test suite for `ubersystem` doesn't include `tabletop`. That's probably why the tests are failing on Travis.

To fix the tests you'll need to add `phonenumbers==7.2.3` to `requirements.txt`. You can use the latest version if you want (`phonenumbers==8.8.1`), but you'll need to also update the version number in the `tabletop` repo.

Hope that helps!

kitsuta commented:
Technically if this gets merged I think we can take out the library from the `tabletop` plugin altogether!

floraline commented:
Is that because this repo is loaded as a dependency when `tabletop` is tested? Would it be better to remove it from `tabletop` altogether?

Thanks @RobRuana for the explanation!

RobRuana commented:
It wouldn't hurt to remove it from `tabletop` altogether. The only thing to watch out for is if we someday remove it from `ubersystem` and forget to add it back to `tabletop`. But then `tabletop` would break immediately, so we'd catch that in no time.

That being said, I second @kitsuta on removing it from `tabletop`!

RobRuana commented on uber/model_checks.py:
```diff
@@ -118,8 +119,11 @@ def group_money(group):
-    if not s.startswith('+'):
-        return len(re.findall(r'\d', s)) != 10 or re.search(c.SAME_NUMBER_REPEATED, re.sub(r'[^0-9]', '', s))
+    try:
+        parsed = phonenumbers.parse(s, 'US')
```
Forgive me, I am not super familiar with the `phonenumbers` library. Will this correctly handle international numbers? Looks like the old code would consider any number that started with `+` to be international and allow it. But what we really want is to actually validate international numbers.

EliAndrewC replied:
The second parameter is simply the default country to assume if no ``+`` is present, e.g. if I give it something like ``703-451-2125``.  The function will treat any number with a leading ``+`` as an international number and do the right thing based on the country code.

RobRuana replied:
Nice!

floraline replied:
My understanding is that the second parameter is more like a fallback option. When a leading `+` is provided, then that means the country code was also provided, so it can validate accordingly. Otherwise, there is no way to know what country the number belongs to, so the second parameter instructs it to assume it is a US number.

floraline commented:
I'm not sure whether we should use `is_valid_number`. If it isn't completely accurate, then it might deny numbers that really are correct. It's a lot easier to say "that looks like a phone number" than it is to say "that is a valid phone number within x locale."

As long as we are doing some basic error-checking, then I think we have done our due diligence. It's up to the user to provide a real number, we are basically just helping them avoid typos.

RobRuana commented:
Fair enough! Even with `is_possible_number`, we're still doing a much better job than we were before!

floraline commented:
@RobRuana It looks like the function I want to test, `_invalid_phone_number`, isn't only used by Attendee model checks. It's also used in five other sideboard plugins. The unit tests for this function don't need to use `Attendee` at all, so is it still okay if I put them in `test_attendee.py`? Or is there another file that would be better?

floraline commented:
Alright, I added a really basic test. I don't usually write unit tests, so please let me know if it can be improved.

In writing the test, I found that `is_possible_number` is perfectly happy with a 7-digit US number. Since I'm pretty sure we always want the area code, I added a condition to `_invalid_phone_number` to account for this. We might be able to use a function like `can_be_internationally_dialled` to check for this kind of thing in international numbers.

RobRuana commented:
Your unit tests look great! There are some neat features in `pytest`, like `@pytest.mark.parametrize`, that might make the tests a little cleaner, but that's mostly syntactic sugar. There are a lot of examples of parametrizing tests in our codebase, like https://github.com/magfest/ubersystem/blob/master/uber/tests/models/test_promo_code.py#L50.

The main benefit of parametrizing a unit test instead of just iterating a list of test data, is that `pytest` will create a separate unit test for each parameter. That makes it easier to tell which specific input in the list is failing.

But again, that's really no big deal. Otherwise, the logic of your unit tests is really great.

The one thing I love about unit tests, is that they really force you to understand exactly what your code is doing. A perfect example is the realization that `is_possible_number` will accept a 7-digit US number. I never would have guessed that!

Thanks for all your contributions. You are a tremendous help!
